invoices/settlement_interceptor.go

```diff
@@ -0,0 +1,244 @@
+package invoices
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnutils"
+)
+
+// SafeCallback is a thread safe wrapper around an InterceptorClientCallback.
+type SafeCallback struct {
+	// mu is a mutex that protects the callback field.
+	mu sync.Mutex
+
+	// callback is the client callback function that is called when an
+	// invoice is intercepted. This function gives the client the ability to
+	// determine how the invoice should be settled.
+	callback InterceptorClientCallback
+}
+
+// Set sets the client callback function.
+func (sc *SafeCallback) Set(callback InterceptorClientCallback) {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+
+	sc.callback = callback
+}
+
+// Exec calls the client callback function.
+func (sc *SafeCallback) Exec(req InterceptClientRequest) error {
```

Comment:
Do we want a separate context or timeout wrapping this callback? IIUC this would block `notifyExitHopHtlcLocked` indefinitely if the interceptor goes offline, crashes, etc.

Reply:
In the latest commits, `Exec` runs in a separate goroutine. And we wait for quit or a client response in `notifyExitHopHtlcLocked`.

There shouldn't be a chance of it blocking if the invoice acceptor not engaged. It's not perfect. But it might do for now.

Reply:
I agree its fine for now & that there's no extra risk if there is no interceptor - I meant if we want to enforce a deadline on the interceptor response. Definitely non-blocking.